src/node_api.cc

```diff
@@ -623,10 +636,42 @@ static void napi_module_register_cb(v8::Local<v8::Object> exports,
       static_cast<const napi_module*>(priv)->nm_register_func);
 }
 
+template <int32_t module_api_version>
+static void node_api_context_register_func(v8::Local<v8::Object> exports,
+                                           v8::Local<v8::Value> module,
+                                           v8::Local<v8::Context> context,
+                                           void* priv) {
+  napi_module_register_by_symbol(
+      exports,
+      module,
+      context,
+      reinterpret_cast<napi_addon_register_func>(priv),
+      module_api_version);
+}
+
+// This function must be augmented for each new Node API version.
+// The key role of this function is to encode module_api_version in the function
+// pointer. We are not going to have many Node API versions and having one
+// function per version is relatively cheap. It avoids dynamic memory
+// allocations or implementing more expensive changes to module registration.
+// Currently AddLinkedBinding is the only user of this function.
+node::addon_context_register_func get_node_api_context_register_func(
+    node::Environment* node_env, int32_t module_api_version) {
+  if (module_api_version <= NODE_API_DEFAULT_MODULE_API_VERSION) {
```

Comment:
Can we add a `static_assert` here so that, if we increment `NAPI_VERSION`, the build will fail if we do not remember to augment this function? The message of the `static_assert` should remind the reader to also update the conditions under which the `static_assert` will fail, because we will likely need to explicitly write the expected `NAPI_VERSION` as a numeric literal. Re https://github.com/nodejs/node/issues/47664.

Reply:
Great idea! Added the `static_assert`.